Add breadcrumbs to help user navigate (fixes #83)

DawoodShahat opened the pull request:
This PR solves [https://github.com/mozilla/glean-dictionary/issues/83](https://github.com/mozilla/glean-dictionary/issues/83)

- Breadcrumb added
- Story for Breadcrumb component added

Here is what it looks like.

![Screenshot 2020-10-14 at 8 02 23 AM](https://user-images.githubusercontent.com/19612755/95938637-b6cc7a00-0df3-11eb-94b9-9d098211f983.png)

DawoodShahat commented:
@wlach One question, I wasn't able to grab table param when navigated to BigQuery table for a particular ping. The table param isn't added to the params object that is passed as a prop to every component. Take a look when possible.

wlach reviewed (changes requested):
This is the right general idea, but I'd like to see a slightly different approach. See comment.

DawoodShahat commented:
Thanks for the insight. I will add the suggestions.

wlach commented on src/App.svelte:
```diff
@@ -11,6 +13,29 @@
+  afterUpdate(() => {
+    const { app, ping, metric } = params;
+
+    links = [{ url: "/", name: "apps" }];
```
I'd bet you can simplify this function a bit (i.e. removing a bunch of the if...else) by using the ternary operator (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator) and array destructuring (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment).

wlach replied:
To be a little more concrete about it, here's what I'd do here:

```js
const links = [
  ...[{ url: "/", name: "apps" }], 
  ...ping ? [{ url: `/apps/${app}/pings/${ping}/`, name: ping }] : [],
  ...metric ? [{ url: `/apps/${app}/metrics/${metric}/`, name: metric }] : [],
];
```

(this example doesn't include BigQuery, leaving that as an exercise for you!)

wlach commented on stories/breadcrumb.stories.js:
```diff
@@ -0,0 +1,37 @@
+
+const AppsLinks = [{ url: "/", name: "apps" }];
+
+const PingLinks = [
```
Again, you could use array destructuring here.

wlach commented on src/App.svelte:
```diff
@@ -11,6 +13,29 @@
+    if (ping) {
+      links.push({
+        url: `/apps/${app}/pings/${ping}`,
+        name: `pings / ${ping}`,
```
I think we can probably skip the `pings /` part of this.

wlach reviewed (changes requested):
This is much closer to what I was expecting, thanks! See comments inline. I think you may have forgotten to add breadcrumbs for the bigquery table pages?

DawoodShahat commented:
Thanks, @wlach I will work on the suggestions. as for the big query page,  I wasn't able to grab table param when navigated to BigQuery table for a particular ping. The table param isn't added to the params object that is passed as a prop to every component. But I will try to come up with the solution.

DawoodShahat commented:
Changes made based on @wlach review:

- Updated ```afterUpdate()``` function to avoid unnecessary conditionals
- Updated route for Big Query table, so that its param is exposed through ```params.bigquery``` 
- Added story for ```TableDetail.svelte``` page

wlach commented on src/App.svelte:
```diff
@@ -20,10 +47,11 @@
 
   page("/", setComponent(AppList));
-  page("/apps/:app/tables/:ping", setComponent(TableDetail));
+  page("/apps/:app/tables/:ping/:bigquery", setComponent(TableDetail));
```
This should actually be `:table` (using `:ping` was actually a mistake)

```suggestion
  page("/apps/:app/tables/:table", setComponent(TableDetail));
```

wlach replied:
Note to do this you'll also need to make some (very minor) changes to the table component: https://github.com/mozilla/glean-dictionary/blob/9990cef3e5e04a06aaeb6d69470516325e5e8535/src/pages/TableDetail.svelte#L6

wlach commented on src/App.svelte:
```diff
@@ -20,10 +47,11 @@
   page("/apps/:app/pings/:ping", setComponent(PingDetail));
   page("/apps/:app/metrics/:metric", setComponent(MetricDetail));
   page("/apps/:app", setComponent(AppDetail));
+  page("*", setComponent(Breadcrumb));
```
Did you mean to add this? It seems like an error and might conflict with some other work that's happening.

DawoodShahat replied:
Oh, my bad. I wanted it to be removed.

wlach commented on src/components/Breadcrumb.svelte:
```diff
@@ -0,0 +1,12 @@
+  export let links;
+</script>
+
+<style>
+</style>
+
```
You can just skip this section if you're not going to define anything.

```suggestion
```

wlach reviewed (changes requested):
This is looking pretty good, code wise! I think it's better if we put the breadcrumbs in the top-left, I actually didn't see them when I first ran your PR. I kind of like the approach I took with Mission Control, which people have generally responded postively to:

![image](https://user-images.githubusercontent.com/20569/96602017-90f61800-12c0-11eb-9121-dc6865b512d1.png)

https://missioncontrol.telemetry.mozilla.org/#/release/windows/main_crashes

(don't worry about reproducing the styling, just the idea is important at this point)

DawoodShahat commented:
Here is a preview

![Screenshot 2020-10-21 at 1 46 40 AM](https://user-images.githubusercontent.com/19612755/96642276-6008f800-133f-11eb-8c67-9522070acb88.png)

wlach commented on src/App.svelte:
```diff
@@ -11,6 +13,36 @@
+            },
+          ]
+        : []),
+    ].filter(Boolean);
```
I am pretty sure this filter should no longer be necessary?

```suggestion
    ]
```

wlach reviewed (changes requested):
This looks great! Just one nit

wlach commented:
Thank you!
